NPB/MG/mg.cpp

```diff
@@ -0,0 +1,1618 @@
+	zero3(u[k], m1[k], m2[k], m3[k]);
+	psinv(r[k], u[k], m1[k], m2[k], m3[k], c, k);
+
+	for (k = lb+1; k <= lt-1; k++) {
+		j = k-1;
+		/*--------------------------------------------------------------------
+		c prolongate from level k-1  to k
+		c-------------------------------------------------------------------*/
+		zero3(u[k], m1[k], m2[k], m3[k]);
+		interp(u[j], m1[j], m2[j], m3[j], u[k], m1[k], m2[k], m3[k], k);
+		/*--------------------------------------------------------------------
+		c compute residual for level k
+		c-------------------------------------------------------------------*/
+		resid(u[k], r[k], r[k], m1[k], m2[k], m3[k], a, k);
+		/*--------------------------------------------------------------------
+		c apply smoother
+		c-------------------------------------------------------------------*/
+		psinv(r[k], u[k], m1[k], m2[k], m3[k], c, k);
+	}
+
+	j = lt - 1;
+	k = lt;
+	interp(u[j], m1[j], m2[j], m3[j], u[lt], n1, n2, n3, k);
+	resid(u[lt], v, r[lt], n1, n2, n3, a, k);
+	psinv(r[lt], u[lt], n1, n2, n3, c, k);
+}
+
+/*--------------------------------------------------------------------
+c-------------------------------------------------------------------*/
+
+static void psinv( dash::NArray<double, 3> &r, dash::NArray<double, 3> &u, int n1, int n2, int n3, double c[4], int k) {
+	/*--------------------------------------------------------------------
+	c-------------------------------------------------------------------*/
+
+	/*--------------------------------------------------------------------
+	c	 psinv applies an approximate inverse as smoother:  u = u + Cr
+	c
+	c	 This  implementation costs  15A + 4M per result, where
+	c	 A and M denote the costs of Addition and Multiplication.
+	c	 Presuming coefficient c(3) is zero (the NPB assumes this,
+	c	 but it is thus not a general case), 2A + 1M may be eliminated,
+	c	 resulting in 13A + 3M.
+	c	 Note that this vectorizes, and is also fine for cache
+	c	 based machines.
+	c-------------------------------------------------------------------*/
+	double r1[n1], r2[n1];
+
+	int myid = dash::myid();
+
+	auto pattern = u.pattern();
+	auto local_beg_gidx = pattern.coords(pattern.global(0));
+  auto local_end_gidx = pattern.coords(pattern.global(pattern.local_size()-1));
+
+	int topcoord = local_beg_gidx[0]-1;
+	int bottomcoord = local_end_gidx[0]+1;
+	int z_ext = u.local.extent(0);
+
+	// double topplane[n2][n1];
+	// double bottomplane[n2][n1];
+	std::vector<double> topplane(n2*n1);
+	std::vector<double> bottomplane(n2*n1);
+
+	dash::Future<double*> fut_top;
+	dash::Future<double*> fut_bot;
+
+	if(topcoord >= 0 && z_ext > 0) {
+		fut_top = dash::copy_async(r.begin()+n2*n1*topcoord, r.begin()+n2*n1*(topcoord+1), &topplane[0]);
+	}
+
+	if(bottomcoord < n3 && z_ext > 0) {
+		fut_bot = dash::copy_async(r.begin()+n2*n1*bottomcoord, r.begin()+n2*n1*(bottomcoord+1), &bottomplane[0]);
+	}
+
+	for(int i3 = 1; i3 < z_ext-1; i3++) {
```

Comment:
If you need performance .local(....) is may be not the best solution. Here you might benefit of plain pointers

Reply:
Here, I would say as well this is a topic which is left to be investigated when someone is using these benchmarks.